### DISCUSSION TITLE
default ipv4 address for an instance

### DISCUSSION THREAD

**Discussion by Romeriz** [bpg/terraform-provider-proxmox#802] [Q&A] [answered]
When I used the Telemate provider I was able to retrieve the IPV4 address assigned to the VM (whether via DHCP or static) and I used that to create my ansible hosts file. I am trying to replicate that.

This is an example :
```hcl
data "template_file" "templs" {
  template = file("./templates/hosts.tpl")
  vars = {
    master_ip = "${join("\n", [for instance in proxmox_vm_qemu.master_vms : join("", [**instance.default_ipv4_address**, " ansible_user=", var.user["name"]])])}"
    node_ip   = "${join("\n", [for instance in proxmox_vm_qemu.slave_vms : join("", [**instance.default_ipv4_address**, " ansible_user=", var.user["name"]])])}"
  }
}
```

Is there some way to replicate this in bgs provider? If not could I request it please

**Comment by bpg-dev** [comment]
I believe you can use `ipv4_addresses` [attribute](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#ipv4_addresses), usage example is [here](https://github.com/bpg/terraform-provider-proxmox/blob/f7e5b5e6d54b1453d4ad47860b2b799f0d048b61/example/resource_virtual_environment_vm.tf#L225). That particular example produces a structure like this:
```
resource_proxmox_virtual_environment_vm_example_ipv4_addresses = tolist([
  tolist([
    "127.0.0.1",
  ]),
  tolist([
    "192.168.3.251",
  ]),
  tolist([]),
])
```
It is basically a list of lists of IP addresses that you can pick the required IP from, depends on your network configuration. 

But make sure the qemu agent is [installed and enabled](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#qemu-guest-agent) on the VM.

**Comment by elias314** [comment, marked as answer]
I found I can use the `ipv4_addresses` attribute on the `proxmox_virtual_environment_vm` resource.  It produces a list of lists which is a little awkward, so I flatten() it, and then use setsubtract() to remove localhost:

```
setsubtract(flatten(proxmox_virtual_environment_vm.control_plane[0].ipv4_addresses), ["127.0.0.1"])
```
